File: mrgeo-mapalgebra/mrgeo-mapalgebra-image/src/main/java/org/mrgeo/mapalgebra/MrsPyramidMapOp.java
/*
 * Copyright 2009-2014 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package org.mrgeo.mapalgebra;

import org.mrgeo.mapreduce.formats.TileClusterInfo;
import org.mrgeo.mapreduce.job.JobCancelledException;
import org.mrgeo.mapreduce.job.JobFailedException;
import org.mrgeo.opimage.MrsPyramidDescriptor;
import org.mrgeo.progress.Progress;
import org.mrgeo.data.DataProviderFactory;
import org.mrgeo.data.DataProviderFactory.AccessMode;
import org.mrgeo.data.image.MrsImageDataProvider;
import org.mrgeo.utils.Bounds;

import java.awt.image.RenderedImage;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;


public class MrsPyramidMapOp extends RasterMapOp
  implements InputsCalculator, BoundsCalculator, TileSizeCalculator,
  MaximumZoomLevelCalculator, TileClusterInfoConsumer
{
  MrsImageDataProvider dp;
  private TileClusterInfo overallTileClusterInfo;

  @Override
  public void addInput(MapOp n) throws IllegalArgumentException
  {
    throw new IllegalArgumentException("MrsPyramidMapOp doesn't take any inputs.");
  }
  
  public static String[] register()
  {
    // returning a "" means don't register this MapOp
    return new String[] { "" };
  }


  @Override
  public void build(Progress p) throws IOException, JobFailedException, JobCancelledException
  {
    if (p != null)
    {
      p.starting();
    }
    //    RenderedImage result = ReplaceableRasterDescriptor.create(pyramidName, pyramid.getBounds());

    //  RenderedImage result = GeographicWritableDescriptor.create(MrsPyramidv1.loadPyramid(new Path(pyramidName))
    //  .getImage(0).getPath().toString());
    //    
    //    if (result.getSampleModel().getDataType() != DataBuffer.TYPE_FLOAT)
    //    {
    //      result = ConvertToFloatDescriptor.create(result);
    //    }
    //    result = TileCacheDescriptor.create(result);
    //  
    //  
    _output = getOutputForHighestLevel();

    if (p != null)
    {
      p.complete();
    }
  }

  @Override
  public void moveOutput(String toName) throws IOException
  {
//    DataProviderFactory.move(getOutputName(), toName, true);
    dp = DataProviderFactory.getMrsImageDataProvider(getOutputName(), AccessMode.READ,
        getProviderProperties());
    if (dp != null)
    {
      dp.move(toName);
    }
//    String sourceName = getOutputName();
//    Configuration conf = HadoopUtils.createConfiguration();
//    Path toPath = new Path(toName);
//    Path sourcePath = new Path(sourceName);
//    FileSystem sourceFs = HadoopFileUtils.getFileSystem(conf, sourcePath);
//    FileSystem destFs = HadoopFileUtils.getFileSystem(conf, toPath);
//    if (!FileUtil.copy(sourceFs, sourcePath, destFs, toPath, false, false, conf))
//    {
//      throw new IOException("Error copying '" + sourceName.toString() +
//          "' to '" + toName.toString() + "'");
//    }
    _outputName = toName;
    _output = getOutputForHighestLevel();
  }

  @Override
  public RenderedImage getRasterOutput() throws IOException
  {
    return getOutputForHighestLevel();
  }

  public RenderedImage getOutputForLevel(int level)
  {
    return MrsPyramidDescriptor.create(dp, level, overallTileClusterInfo,
        getProviderProperties());
  }

  public RenderedImage getOutputForHighestLevel() throws IOException
  {
    return getOutputForLevel(dp.getMetadataReader().read().getMaxZoomLevel());
  }

  @Override
  public String getOutputName()
  {
    return this.dp.getResourceName();
//    return getOutputPathForHighestLevel();
  }

  public void setDataProvider(final MrsImageDataProvider dp)
  {
    this.dp = dp;
  }

  public MrsImageDataProvider getDataProvider()
  {
    return dp;
  }

  @Override
  public Bounds calculateBounds() throws IOException
  {
    if (dp != null)
    {
      return dp.getMetadataReader().read().getBounds();
    }
    return Bounds.world;
  }

  @Override
  public int calculateMaximumZoomlevel() throws IOException
  {
    if (dp != null)
    {
      return dp.getMetadataReader().read().getMaxZoomLevel();
    }
    return 0;
  }
  
  @Override
  public int calculateTileSize() throws IOException
  {
    return dp.getMetadataReader().read().getTilesize();
  }

  @Override
  public Set<String> calculateInputs()
  {
    Set<String> inputPyramids = new HashSet<String>();
    inputPyramids.add(dp.getResourceName());

    return inputPyramids;
  }

  @Override
  public void setOverallTileClusterInfo(TileClusterInfo tileClusterInfo)
  {
    this.overallTileClusterInfo = tileClusterInfo;
  }

  @Override
  public MrsPyramidMapOp clone()
  {
    MrsPyramidMapOp result = (MrsPyramidMapOp) super.clone();
    result.setDataProvider(dp);
    return result;
  }

  @Override
  public String toString()
  {
    if (dp != null)
    {
      return "[" + dp.getResourceName() + "]";
    }
    return "[unknown]";
  }

}

File: mrgeo-core/src/main/scala/org/mrgeo/mapalgebra/raster/MrsPyramidMapOp.scala
/*
 * Copyright 2009-2016 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 *
 */

package org.mrgeo.mapalgebra.raster

import java.io.IOException

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.apache.spark.{SparkConf, SparkContext}
import org.mrgeo.data.image.MrsImageDataProvider
import org.mrgeo.data.rdd.RasterRDD
import org.mrgeo.image.MrsPyramidMetadata
import org.mrgeo.job.JobArguments
import org.mrgeo.mapalgebra.MapOp
import org.mrgeo.utils.SparkUtils
import org.mrgeo.utils.tms.Bounds

object MrsPyramidMapOp {
  def apply(dataprovider: MrsImageDataProvider) = {
    new MrsPyramidMapOp(dataprovider)
  }

  def apply(mapop:MapOp):Option[MrsPyramidMapOp] = {
    mapop match {
    case rmo:MrsPyramidMapOp => Some(rmo)
    case _ => None
    }
  }
}

@SuppressFBWarnings(value = Array("CN_IMPLEMENTS_CLONE_BUT_NOT_CLONEABLE"), justification = "_Does_ implement Cloneable()")
class MrsPyramidMapOp private[raster] (dataprovider: MrsImageDataProvider)
    extends RasterMapOp with Cloneable {
  private var rasterRDD:Option[RasterRDD] = None
  private var zoomForRDD: Option[Int] = None
  private var maxZoomForRDD: Option[Int] = None
  private var bounds: Option[Bounds] = None
  private var mapOpForBounds: Option[RasterMapOp] = None

  override def rdd(zoom:Int):Option[RasterRDD]  = {
    load(zoom)
    rasterRDD
  }

  def rdd():Option[RasterRDD] = {
    load()
    rasterRDD
  }

  @SuppressFBWarnings(value = Array("CN_IDIOM_NO_SUPER_CALL"), justification = "No need to call super.clone()")
  override def clone = MrsPyramidMapOp(dataprovider)

  private def getBounds: Option[Bounds] = {
    mapOpForBounds match {
      case Some(op) =>
        Some(op.metadata().getOrElse(
          throw new IOException("Unable to get metadata for the bounds raster")).getBounds)
      case None => bounds
    }
  }

  private def load(zoom:Int = -1)  = {

    if (context == null) {
      throw new IOException("Error creating RasterRDD, can not create an RDD without a SparkContext")
    }

    // If we haven't loaded anything yet
    if (rasterRDD.isEmpty || zoomForRDD.isEmpty || maxZoomForRDD.isEmpty) {
      val data = if (zoom <= 0) {
        getBounds match {
          case Some(b) => SparkUtils.loadMrsPyramidAndMetadata(dataprovider, b, context())
          case None => SparkUtils.loadMrsPyramidAndMetadata(dataprovider, context())
        }
      }
      else {
        getBounds match {
          case Some(b) => SparkUtils.loadMrsPyramidAndMetadata(dataprovider, zoom, b, context())
          case None => SparkUtils.loadMrsPyramidAndMetadata(dataprovider, zoom, context())
        }
      }

      metadata(data._2)
      rasterRDD = Some(data._1)
      maxZoomForRDD = Some(data._2.getMaxZoomLevel)
      zoomForRDD = Some(if (zoom > 0) zoom else data._2.getMaxZoomLevel)
    }
    // if we sent in a zoom and it is different than the current loaded one
    else if (zoom > 0 && zoom != zoomForRDD.get) {
      rasterRDD = Some(
        getBounds match {
          case Some(b) => SparkUtils.loadMrsPyramid(dataprovider, zoom, b, context())
          case None => SparkUtils.loadMrsPyramid(dataprovider, zoom, context())
        }
      )
      zoomForRDD = Some(zoom)
    }
    // if we didn't pass a zoom and it is not max zoom
    else if (zoomForRDD.get != maxZoomForRDD.get) {
      rasterRDD = Some(
        getBounds match {
          case Some(b) => SparkUtils.loadMrsPyramid(dataprovider, maxZoomForRDD.get, b, context())
          case None => SparkUtils.loadMrsPyramid(dataprovider, maxZoomForRDD.get, context())
        })
      zoomForRDD = Some(maxZoomForRDD.get)
    }
  }

  def setBounds(bounds: Bounds): Unit = {
    this.bounds = Some(bounds)
    rasterRDD = None
  }

  def setBounds(mapOpForBounds: RasterMapOp): Unit = {
    this.mapOpForBounds = Some(mapOpForBounds)
    rasterRDD = None
  }

  def zoom():Int = {
    zoomForRDD.get
  }

  override def metadata():Option[MrsPyramidMetadata] =  {
    load()
    getBounds match {
      case Some(b) =>
        val meta = super.metadata()
        meta match {
          case Some(m) =>
            m.setBounds(b)
            Some(m)
          case None => meta
        }
      case None => super.metadata()
    }
  }

  // nothing to do here...
  override def setup(job: JobArguments, conf: SparkConf): Boolean = true
  override def execute(context: SparkContext): Boolean = true
  override def teardown(job: JobArguments, conf: SparkConf): Boolean = true

}
